viewer/src/culling/TakenSectorTree.ts

```diff
@@ -0,0 +1,123 @@
+import { traverseUpwards, traverseDepthFirst } from '../utils/traversal';
+import { PrioritizedWantedSector, DetermineSectorCostDelegate } from './types';
+
+export class TakenSectorTree {
+  get totalCost(): number {
+    return this._totalCost;
+  }
+  private readonly sectors: {
+    sector: SectorMetadata;
+    priority: number;
+    cost: number;
+    lod: LevelOfDetail;
+  }[] = [];
+  private readonly determineSectorCost: DetermineSectorCostDelegate;
+
+  private _totalCost = 0.0;
+
+  constructor(sectorRoot: SectorMetadata, determineSectorCost: DetermineSectorCostDelegate) {
+    this.determineSectorCost = determineSectorCost;
+    // Allocate space for all sectors
+    traverseDepthFirst(sectorRoot, x => {
+      this.sectors.length = Math.max(this.sectors.length, x.id);
+      this.sectors[x.id] = { sector: x, priority: -1, cost: 0, lod: LevelOfDetail.Discarded };
+      return true;
+    });
+    this.setSectorLod(sectorRoot.id, LevelOfDetail.Simple);
+  }
+
+  getWantedSectorCount(): number {
+    return this.sectors.reduce((count, x) => {
+      count = x.lod !== LevelOfDetail.Discarded ? count + 1 : count;
+      return count;
+    }, 0);
+  }
+
+  toWantedSectors(scene: SectorScene): PrioritizedWantedSector[] {
+    return this.sectors
+      .map((x, id) => {
+        const wanted: PrioritizedWantedSector = {
+          sectorId: id,
+          levelOfDetail: x.lod,
+          metadata: x.sector,
+          priority: x.priority,
+          scene
+        };
+        return wanted;
+      })
+      .sort((l, r) => r.priority - l.priority);
+  }
+
+  markSectorDetailed(sectorId: number, priority: number) {
+    this.setSectorPriority(sectorId, priority);
+    if (this.sectors[sectorId].lod === LevelOfDetail.Detailed) {
+      return;
+    }
+    traverseUpwards(this.sectors[sectorId].sector, x => {
+      switch (this.sectors[x.id].lod) {
+        case LevelOfDetail.Simple:
+          this.replaceSimpleWithDetailed(x.id);
+          break;
+        case LevelOfDetail.Discarded:
+          this.setSectorLod(x.id, LevelOfDetail.Detailed);
+          break;
+        default:
+        // Already detailed
+      }
+      return true;
+    });
+
+    this.markAllDiscardedChildrenAsSimple(sectorId);
+  }
+
+  /**
+   * Replaces a Simple sector by marking all its non-detailed children as Simple and
+   * marking the sector itself Detailed.
+   */
+  private replaceSimpleWithDetailed(sectorId: number) {
+    assert(
+      this.sectors[sectorId].lod === LevelOfDetail.Simple,
+      `Sector ${sectorId} must be a Simple-sector, but got ${this.sectors[sectorId].lod}`
+    );
+    this.setSectorLod(sectorId, LevelOfDetail.Detailed);
+    this.markAllDiscardedChildrenAsSimple(sectorId);
+  }
+
+  private markAllDiscardedChildrenAsSimple(sectorId: number) {
+    for (const child of this.sectors[sectorId].sector.children) {
+      if (this.getSectorLod(child.id) === LevelOfDetail.Discarded) {
+        if (child.facesFile.fileName !== null) {
```

Comment:
If the filename is null, should we not stop here and just use the empty sector as a simple representation rather than loading child sectors with simple data with higher number of faces?

Reply:
Good point, fixed.